var/spack/repos/builtin/packages/chapel/package.py

```diff
@@ -3,24 +3,784 @@
+            raise InstallError(
+                "Chapel did not recognize the {0} compiler".format(self.spec.compiler.name)
+            )
+        env.set("CHPL_HOST_COMPILER", self.compiler_map[self.spec.compiler.name])
+        env.set("CHPL_TARGET_COMPILER", self.compiler_map[self.spec.compiler.name])
+
+        # Undo spack compiler wrappers:
+        # the C/C++ compilers must work post-install
+        if is_CrayEX() and os.environ.get("CRAYPE_DIR"):
+            real_cc = join_path(os.environ["CRAYPE_DIR"], "bin", "cc")
+            real_cxx = join_path(os.environ["CRAYPE_DIR"], "bin", "CC")
+        else:
+            real_cc = self.compiler.cc
+            real_cxx = self.compiler.cxx
+        env.set("CHPL_TARGET_CC", real_cc)
+        env.set("CHPL_TARGET_CXX", real_cxx)
+
+    def setup_chpl_comm(self, env, spec):
+        env.set("CHPL_COMM", spec.variants["comm"].value)
+
+    def setup_chpl_llvm(self, env):
+        if self.spec.variants["llvm"].value == "spack":
+            env.set(
+                "CHPL_LLVM_CONFIG", "{0}/{1}".format(self.spec["llvm"].prefix, "bin/llvm-config")
+            )
+            env.set("CHPL_LLVM", "system")
+        else:
+            env.set("CHPL_LLVM", self.spec.variants["llvm"].value)
+
+    def setup_if_not_unset(self, env, var, value):
+        if value != "unset":
+            if value == "spack":
+                value = "system"
+            env.set(var, value)
+
+    def setup_env_vars(self, env):
+        for v in self.spec.variants.keys():
+            self.setup_if_not_unset(env, "CHPL_" + v.upper(), self.spec.variants[v].value)
+        self.setup_chpl_llvm(env)
+        self.setup_chpl_compilers(env)
+        self.setup_chpl_platform(env)
+
+        # TODO: a function to set defaults for things where we removed variants
+        # We'll set to GPU later if +rocm or +cuda requested
+        env.set("CHPL_LOCALE_MODEL", "flat")
+
+        if self.spec.variants["gmp"].value == "spack":
+            env.set("CHPL_GMP", "system")
+            # TODO: why must we add to CPATH to find gmp.h
+            # TODO: why must we add to LIBRARY_PATH to find lgmp
+            env.prepend_path("CPATH", self.spec["gmp"].prefix.include)
+            env.prepend_path("LIBRARY_PATH", self.spec["gmp"].prefix.lib)
+        else:
+            env.set("CHPL_GMP", self.spec.variants["gmp"].value)
+
+        if "yaml" in self.get_package_modules:
+            env.prepend_path("PKG_CONFIG_PATH", self.spec["libyaml"].prefix.lib.pkgconfig)
+            env.prepend_path("CPATH", self.spec["libyaml"].prefix.include)
+
+        if "zmq" in self.get_package_modules:
+            env.prepend_path("CPATH", self.spec["libzmq"].prefix.include)
+            env.prepend_path("LD_LIBRARY_PATH", self.spec["libzmq"].prefix.lib)
+            env.prepend_path("PKG_CONFIG_PATH", self.spec["libzmq"].prefix.lib.pkgconfig)
+            env.prepend_path("PKG_CONFIG_PATH", self.spec["libsodium"].prefix.lib.pkgconfig)
```

Comment:
all of these changes (i.e. those touching env pathing vars) were made after running into problems when the additions were not present. After some discussion with @bonachea last night, we believe there's a challenge to make the chapel language and compiler work as a spack package because the dependencies that spack supplies are relevant to not only chapel, but to the applications that chapel builds. Currently, we do not encode `RPATH`s for all the dependencies in the executable and so when these compiled binaries try to run, they lack the necessary linking information.

Reply:
To restate @arezaii 's point more directly: 
The chapel package we are building is itself a compiler, and long after `spack install` creates the `chpl` binary, end users "use" Chapel by invoking the `chpl` compiler to compile and link their applications written in the Chapel language. 

The execution of the `chpl` compiler requires calls to the system C compiler and linker, which cannot automatically benefit from the spack compiler wrappers or other spack magic because these invocations happen long after `spack install` is complete. Also, we believe the `chpl` compiler does not (yet) correctly pass RPATHs to the application link step corresponding to libraries built using Spack; the `chpl` application link step instead mistakenly assumes that any such libraries (detected as pre-existing "system" libraries during Chapel configure) are part of the default system linker search path. We hope to eventually improve logic in the Chapel compiler to resolve that latter misbehavior, but that's not a trivial change and won't help the current releases of Chapel.


The insertion of all these `LD_LIBRARY_PATH` settings in `setup_run_environment()` primarily ensures that end users can `spack load chapel` (or load a Spack-generated module) to get the necessary library paths to run their Chapel-built applications. We recognize this is a non-ideal user experience, as ideally end users of a Chapel install should never need to know or care that Spack was used by their sysadmin to install Chapel. However it provides a working solution today for current Chapel releases until the Chapel compiler can be improved to integrate better with Spack-provided dependencies. 

I'm less confident that the `CPATH` and `PKG_CONFIG_PATH` are strictly necessary in `setup_run_environment()`, but they also don't seem harmful.